core/vdbe/mod.rs

```diff
@@ -4286,6 +4302,70 @@ fn exec_if(reg: &OwnedValue, jump_if_null: bool, not: bool) -> bool {
     }
 }
 
+fn apply_affinity_char(target: &mut Register, ch: u8) {
+    if let Register::OwnedValue(value) = target {
+        if matches!(value, OwnedValue::Blob(_)) {
+            return;
+        }
+        match ch {
+            /* BLOB */
+            0x41 => return,
+            /* TEXT */
+            0x42 => {
+                if matches!(value, OwnedValue::Text(_) | OwnedValue::Null) {
+                    return;
+                }
+                let text = value.to_string();
+                *value = OwnedValue::from_text(&text);
```

Comment:
This to string is an allocation and then another one with OwnedValue. Can't we just have a new from method that owns the string argument?